pg_search/src/postgres/utils.rs

```diff
@@ -65,74 +65,133 @@ pub fn u64_to_item_pointer(value: u64, tid: &mut pg_sys::ItemPointerData) {
-        let isnull = *isnull.add(attno);
+            if is_array {
+                (MergeStrategy::Array, search_field.id, datum, base_oid)
+            } else if is_json {
+                (MergeStrategy::Json, search_field.id, datum, base_oid)
+            } else {
+                (MergeStrategy::Field, search_field.id, datum, base_oid)
+            }
+        })
+        .partition(|(strategy, _, _, _)| matches!(strategy, MergeStrategy::Json));
 
-        let SearchFieldName(key_field_name) = schema.key_field().name;
-        if key_field_name == attname && isnull {
-            return Err(IndexError::KeyIdNull(key_field_name));
-        }
+    if !other_fields
+        .iter() // Check if key field was filtered out for being null.
+        .any(|(_, id, _, _)| id == &schema.key_field().id)
+    {
+        return Err(IndexError::KeyIdNull);
+    }
 
-        if isnull {
-            continue;
-        }
+    let json_tantivy_values_for_columns = json_fields
+        .into_iter()
+        .map(|(_, field_id, datum, base_oid)| {
+            match TantivyValue::try_from_datum_json(datum, base_oid) {
+                Ok(iter) => iter
+                    .map(|value| (field_id, value.tantivy_schema_value()))
+                    .collect(),
+                Err(err) => panic!("error processing json data: {err}"),
+            }
+        })
+        .collect::<Vec<Vec<_>>>();
+
+    let mut documents: Vec<SearchDocument> = json_tantivy_values_for_columns
+        .iter()
+        .enumerate()
+        .flat_map(|(current_column_index, json_values_for_column)| {
+            json_values_for_column
+                .iter()
+                .map(move |(field_id, field_value)| {
+                    let mut document = schema.new_document();
+                    document.insert(*field_id, field_value.clone());
+                    (document, current_column_index)
+                })
+        })
+        .map(|(mut document, skip_index)| {
+            for (index, column_json_values) in json_tantivy_values_for_columns.iter().enumerate() {
+                if index != skip_index {
+                    for (other_field_id, other_field_value) in column_json_values {
+                        document.insert(*other_field_id, other_field_value.clone());
+                    }
+                }
+            }
+            document
+        })
+        .collect();
+
+    // If there were no JSON fields, make sure we have at least one document.
+    if documents.is_empty() {
+        documents.push(schema.new_document());
+    }
 
-        if is_array {
-            for value in TantivyValue::try_from_datum_array(datum, base_oid)? {
-                document.insert(search_field.id, value.tantivy_schema_value());
+    // Insert ctid values into all documents
+    for document in &mut documents {
+        let ctid_index_value = item_pointer_to_u64(ctid);
+        document.insert(schema.ctid_field().id, ctid_index_value.into())
+    }
```

Comment:
Rather than looping over the set of documents again, we should do this inline, up above, as the documents are otherwise being built.

Also, we can do `let ctid_index_value = item_pointer_to_u64(ctid);` once, at the start of the function.  Micro-optimization but there's no need in doing it N times.